Follow up to PR #38527

rallytime opened the pull request:
### What does this PR do?
This PR back-ports the changes made by @rbjorklin in #38527 to the 2016.3 branch and also adds back in the `log_file` and `pidfile`value settings.

This ensures that if `rest_timeout` values are set in the master config file, the value defined in the `DEFAULT_API_OPTS` will not overwrite the value defined in the master config file. (This will also be the case for any additions API default opts that get added to `config/__init__.py` down the line.)

This change also ensures that that the `log_file` and `pidfile` values *are* overwritten by the values specified by `api_logfile` and `api_pidfile`, respectively, either explicitly by the user in the master config file or by the values defined by `DEFAULT_API_OPTS`.

I would like to write a test case for this as well, but I wanted to get some feedback on this approach first.

This approach should be covering both the `rest_timeout` setting bug as well as the `log_file` bugs that have been filed recently against this area of the code. The relevant PRs and bug reports are linked below. A more detailed discussion of this change is in #38527.

I would love it if @whiteinge could review this change as well.

### What issues does this PR fix or reference?
PRs:
- #38527
- #38560
- #38570

Issues:
- #38479
- #38524

Commit:
salt-api no longer forces the default timeout

Conflicts:
  - salt/config/__init__.py

Commit:
Make sure the pidfile and log_file values are overriden by api opts

whiteinge commented on salt/config/__init__.py:
```diff
@@ -3315,15 +3315,18 @@ def api_config(path):
+    # Let's set the pidfile and log_file values in opts to api settings
+    opts.update({
         'pidfile': opts.get('api_pidfile', DEFAULT_API_OPTS['api_pidfile']),
         'log_file': opts.get('api_logfile', DEFAULT_API_OPTS['api_logfile']),
```
I looked through the history for this file and I do not understand the purpose of setting `pidfile` and `log_file` to the corresponding `api_*` values **at all**. Shot in the dark: I wonder if the daemon that manages netapi modules is expecting those values and should be switched to look for the `api_*` variants.

That said, in order to correctly work with a non-default `root_dir` setting both those values would also need to be added to the `prepend_root_dir` call below.

I think this PR looks good with that change in place (ignoring the `pidfile`/`log_file` unknowns).

rallytime replied:
@whiteinge Hm. Yeah I am assuming there was something going on later down the line that is using those `pidfile` and `log_file` settings somewhere in salt-api, especially since there was a change in this code recently in #37272, which removed the `log_file` setting and caused issue #38479. I didn't want to cause that bug again since that was fixed recently on this branch.

Just to make sure I am following you correctly, you're saying that the `prepend_root_dir` option also needs the `pidfile` and `log_file` settings? Something like this?
```
prepend_root_dir(opts, [
          'api_pidfile',
          'api_logfile',
     ])
```
Or am I misunderstanding what you said?

whiteinge replied:
No, you're exactly right. That addition makes everything work correctly for me in a quick local smoke test.

rallytime replied:
Perfect. I have made that change locally, but I am also writing some quick unit tests for this change. I'll push everything up when those are finished.

cachedout replied:
@rallytime Sounds good. Ping me when ready.

Commit:
Make sure the pidfile and log_file values are overriden by api opts

Commit:
Add some simple unit tests for salt.config.api_config function

This will help prevent regressions in this area as discussed in
PR #38527.

rallytime commented:
OK @cachedout and @whiteinge this should be good to go now, depending on the test run.

Commit:
Test clean up

rallytime commented:
Hrm. I am not sure how that reference to the `/hello/word` mocked log file settings is persisting and then getting consumed in the parsers unit tests...

Commit:
Keep a copy of the DEFAULT_API_OPTS and restore them after the test run

Without this copy, the settings in DEFAULT_API_OPTS was getting overridden
by the new settings in the unit test. Let's copy those default settings,
and then restore them at the end of the test. These tests have also been
marked as destructive.

rallytime commented:
Found where those were getting set (in the salt.config.api_config function call itself). I have adjusted the tests to make a copy of those defaults before the test runs and then also restored those default settings at the end of the test run.

rallytime commented:
Go Go Jenkins!

rallytime commented:
Go Go Jenkins!

Commit:
Pylint fix: add line at end of file
